grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java

```diff
@@ -571,10 +578,28 @@ private HttpResponse serve0(ServiceRequestContext ctx, HttpRequest req,
                 } else {
                     try {
                         ctx.setAttr(FramedGrpcService.RESOLVED_GRPC_METHOD, spec.method);
-                        // Set JSON media type (https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/grpc_json_transcoder_filter#sending-arbitrary-content)
+                        final MediaType mediaType;
+                        final HttpData content;
+
+                        // https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/grpc_json_transcoder_filter#sending-arbitrary-content
+                        if (HttpBody.getDescriptor().equals(spec.methodDescriptor.getInputType())) {
+                            // TODO: Determine media type dynamically from content.
+                            mediaType = ctx.request().contentType();
+                            content = convertToArbitraryHttpData(ctx, clientRequest);
+                        } else {
+                            mediaType = MediaType.JSON_UTF_8;
+                            content = convertToJson(ctx, clientRequest, spec);
+                        }
+
+                        // Media type can only be nullif the request doesn't have a content type.
+                        if (mediaType == null) {
+                            throw new IllegalArgumentException("Unsupported content-type: " +
```

Comment:
Maybe should we fall back to octet-stream?

Reply:
First time hearing about it, mozilla docs also point to the same direction, https://arc.net/l/quote/dteigvyk

> Generic binary data (or binary data whose true type is unknown) is application/octet-stream.

I think it makes sense 👍